src/libraries/System.Net.Quic/src/System/Net/Quic/NetEventSource.Quic.Counters.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Diagnostics.Metrics;
+using System.Net.Quic;
+
+using Microsoft.Quic;
+using static Microsoft.Quic.MsQuic;
+
+namespace System.Net
+{
+    internal sealed partial class NetEventSource
+    {
+        private static Meter s_meter = new Meter("Private.InternalDiagnostics.System.Net.Quic");
+        private static readonly long[] s_counters = new long[(int)QUIC_PERFORMANCE_COUNTERS.MAX];
+
+        public static readonly ObservableGauge<long> MsQuicCountersGauge = s_meter.CreateObservableGauge<long>(
```

Comment:
Shouldn't this be counter/instrument per MsQuic counter? Is this like one .NET instrument with many MsQuic counters in it, just with different tags? How does this even look like when observed in `dotnet counters` etc.?

Reply:
for `dotnet counters` example, see PR description.

I would prefer having separate instruments/counters per MsQuic counter, but as explained in PR description, there is no easy way to implement that effectively.

We can only get all counters at once (one GetParam call which copies the array of 30 values), and if we have instruments for each MsQuic counter, with each one invoking the get-current-value-callback, we either
- fetch the fresh counter values array from each callback just to read a single value.
- figure out how to refresh only when necessary, which is problematic, because
  - we don't know the frequency of fetching (the monitoring tool dictates that)
  - individual counters can be disabled/filtered out by the tool, so we don't have a guarantee that e.g. CONN_ACTIVE will be always fetched first

Reply:
> figure out how to refresh only when necessary

For example, timestamp the last fetch and do not fetch more that once per 0.5 second.

But thinking about this more, this is mostly for us, it's not necessary to have this nice and proper. But if we leave it like this, could you please summarize the reasoning behind this decision in a comment.

Reply:
> For example, timestamp the last fetch and do not fetch more that once per 0.5 second.

This should work fine. And it might actually be better we do it this way, I tried to play with Prometheus exporting via `dotnet-monitor` and the previous version did no get exported.